File: pypi/trackfw/commands/update.py
"""
commands/update.py — trackfw update (Python CLI).

Scope: the current repository only (docs/cli-parity.md, "`trackfw update`
vs `trackfw update harness`"). This command never mutates global state —
every write below is rooted at `cwd`, and the Codex integration block below
plans/applies with `scope="project"` explicitly. `trackfw update harness`
(trackfw/commands/update_harness.py) is the counterpart that refreshes the
user's global harness (`~/.claude`, `~/.codex`, etc.) and runs from
anywhere, without a `trackfw.yaml`.

Escopo reduzido: atualiza somente as regras de agente (blocos marker-delimited),
os agent hooks e a integração Codex de projeto já instalada. Gates (hooks/CI)
e Claude commands requerem o CLI Go ou Node.js — ver docs/cli-parity.md.

--dry-run, --json, --targets and --install-missing (added for
REQ-2026-07-29-barrier-governanca-e-autoridade-do-orquestrador, ML-6F) report
the same four-state model (updated/skipped/missing/failed) as `trackfw update
harness`, over a "scope": "project" JSON document — see docs/cli-parity.md.
The declared PROJECT_TARGET_IDS below is this runtime's reduced surface; it is
NOT required to match Go/Node.js's project target list byte-for-byte — only
the states, flags and JSON document shape are shared (see the docstring on
"Escopo reduzido" above, which the Go/Node.js CLIs do not implement).
"""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import shutil
import tempfile
from pathlib import Path
from typing import Any

from trackfw.commands import update_harness
from trackfw.commands.update_harness import (
    STATE_FAILED,
    STATE_MISSING,
    STATE_SKIPPED,
    STATE_UPDATED,
)

AGENT_RULES_RELATIVE_PATHS = [
    "CLAUDE.md",
    "AGENTS.md",
    "GEMINI.md",
    os.path.join(".github", "copilot-instructions.md"),
    ".windsurfrules",
    os.path.join(".amazonq", "developer", "guidelines.md"),
    os.path.join(".cursor", "rules", "trackfw.mdc"),
]

AGENT_HOOKS_RELATIVE_PATHS = [
    os.path.join(".claude", "settings.json"),
    os.path.join(".codex", "hooks.json"),
    os.path.join(".gemini", "settings.json"),
    os.path.join(".kiro", "hooks", "trackfw-attention.json"),
    os.path.join(".github", "hooks", "trackfw-attention.json"),
    os.path.join(".cursor", "hooks.json"),
    os.path.join("scripts", "trackfw-attention-signal.sh"),
    os.path.join("scripts", "trackfw-attention-cleanup.sh"),
]

CODEX_PROJECT_AGENTS_DISPLAY_PATH = ".codex/agents, .agents/skills"

# PROJECT_TARGET_IDS — this runtime's declared, fixed-order project-scope
# target list (see module docstring: reduced scope, documented exception).
PROJECT_TARGET_IDS = ["agent-rules", "agent-hooks", "codex-project-agents"]


def register(subparsers: argparse.ArgumentParser) -> None:
    parser = subparsers.add_parser(
        "update",
        help="Update trackfw rules in agent config files (agent rules only)",
    )
    parser.add_argument("--dry-run", action="store_true", help="Compute and report states without writing anything")
    parser.add_argument("--json", action="store_true", help="Emit the result document instead of the text report")
    parser.add_argument("--targets", help="Comma-separated subset of target ids")
    parser.add_argument(
        "--install-missing",
        action="store_true",
        help="Allow missing targets to be installed instead of merely reported",
    )
    # `update_action` is optional (required=False, the argparse default) so
    # bare `trackfw update` keeps running `_run` below via `set_defaults`.
    # Only when the user types `trackfw update harness` does the child
    # parser's own `set_defaults(func=...)` override it.
    update_actions = parser.add_subparsers(dest="update_action")
    update_harness.register(update_actions)
    parser.set_defaults(func=_dispatch)


def _dispatch(args: argparse.Namespace) -> None:
    if args.dry_run or args.json or args.targets or args.install_missing:
        _run_project(args)
        return
    _run(args)


def _run(args: argparse.Namespace) -> None:
    cwd = os.getcwd()
    yaml_path = os.path.join(cwd, "trackfw.yaml")

    if not os.path.exists(yaml_path):
        print("Erro: trackfw.yaml não encontrado — execute trackfw init primeiro")
        raise SystemExit(1)

    print("trackfw update — atualizando regras de agente...\n")

    from trackfw.generators.init_gen import inject_rules_detected
    try:
        inject_rules_detected(cwd)
        print("  Regras de agente atualizadas (CLAUDE.md, GEMINI.md, etc.)")
    except Exception as e:
        print(f"  Aviso: falha ao atualizar regras: {e}")

    from trackfw.generators.hooks import inject_hooks_detected
    try:
        inject_hooks_detected(cwd)
        print('  ✓ agent hooks atualizados')
    except Exception as e:
        print(f'  ⚠ agent hooks: {e}')

    if os.path.exists(os.path.join(cwd, "AGENTS.md")) or os.path.isdir(os.path.join(cwd, ".codex")):
        from trackfw import identity
        from trackfw.identity import IdentityError

        # Identity errors must abort the command — never fall back silently
        # to the neutral default, which would revert the user's identity.
        try:
            ident = identity.load(os.path.expanduser("~"))
        except IdentityError as e:
            print(f"update: identidade invalida: {e}")
            raise SystemExit(2) from e

        try:
            from trackfw.integrations.catalog import plan_deployments
            from trackfw.integrations.manager import IntegrationManager
            manager = IntegrationManager(cwd)
            _, plans = plan_deployments("agents", target_ids=["codex"], scope="project", identity_cfg=ident)
            plans = [plan for plan, status in zip(plans, manager.list(plans)) if status["state"] != "not-installed"]
            manager.update(plans)
            _, plans = plan_deployments("skills", target_ids=["codex"], scope="project", identity_cfg=ident)
            plans = [plan for plan, status in zip(plans, manager.list(plans)) if status["state"] != "not-installed"]
            manager.update(plans)
        except Exception as e:
            print(f"  ⚠ Codex integration: {e}")

    print()
    print("  Nota: este CLI Python atualiza apenas as regras de agente.")
    print("  Para atualizar gates (hooks/CI) e Claude commands, use:")
    print("    trackfw update   (CLI Go)")
    print("    npx trackfw update   (CLI Node.js)")

    print("\ntrackfw update concluído")
    try:
        from trackfw.generators.init_gen import print_architect_next_steps
        print_architect_next_steps(cwd)
    except Exception:
        pass


# ---------------------------------------------------------------------------
# --dry-run / --json / --targets / --install-missing — four-state model,
# mirroring internal/generators/update.go's runFileTarget and
# npm/src/lib/update-engine.js's runFileTarget for this runtime's reduced
# project-scope target set.
# ---------------------------------------------------------------------------


def _hash_path(path: str) -> str | None:
    """Returns None when path does not exist, a content hash for a file, or
    a hash of the recursive (relative-path, content-hash) listing for a
    directory."""
    if not os.path.exists(path):
        return None
    if os.path.isfile(path):
        return hashlib.sha256(Path(path).read_bytes()).hexdigest()
    entries = []
    for root, _dirs, files in os.walk(path):
        for name in files:
            full = os.path.join(root, name)
            rel = os.path.relpath(full, path)
            entries.append(f"{rel}:{hashlib.sha256(Path(full).read_bytes()).hexdigest()}")
    entries.sort()
    return hashlib.sha256("\n".join(entries).encode("utf-8")).hexdigest()


def _hash_rel_paths(root: str, rel_paths: list[str]) -> list[str | None]:
    return [_hash_path(os.path.join(root, rel)) for rel in rel_paths]


def _all_missing(hashes: list[str | None]) -> bool:
    return all(h is None for h in hashes)


def _run_file_target(
    target_id: str,
    display_path: str,
    root: str,
    rel_paths: list[str],
    apply,
    dry_run: bool,
    install_missing: bool,
) -> dict[str, Any]:
    """Computes updated/skipped/missing/failed for a target whose only
    observable effect is writing under rel_paths (relative to root), by
    diffing content hashes before/after invoking apply(root). "missing"
    never installs: apply is never called when every rel_path is absent and
    install_missing is not set."""
    before = _hash_rel_paths(root, rel_paths)
    if _all_missing(before) and not install_missing:
        return {"id": target_id, "state": STATE_MISSING, "path": display_path}

    try:
        apply(root)
    except Exception as error:
        return {"id": target_id, "state": STATE_FAILED, "path": display_path, "message": str(error)}

    after = _hash_rel_paths(root, rel_paths)
    if _all_missing(before) and _all_missing(after):
        return {"id": target_id, "state": STATE_MISSING, "path": display_path}
    if before == after:
        return {"id": target_id, "state": STATE_SKIPPED, "path": display_path}
    return {"id": target_id, "state": STATE_UPDATED, "path": display_path}


def _codex_project_agents_target(root: str, dry_run: bool, install_missing: bool) -> dict[str, Any]:
    detected = os.path.exists(os.path.join(root, "AGENTS.md")) or os.path.isdir(os.path.join(root, ".codex"))
    if not detected:
        return {"id": "codex-project-agents", "state": STATE_MISSING, "path": CODEX_PROJECT_AGENTS_DISPLAY_PATH}

    try:
        from trackfw import identity
        from trackfw.identity import IdentityError
        from trackfw.integrations.catalog import plan_deployments
        from trackfw.integrations.manager import IntegrationManager

        try:
            ident = identity.load(os.path.expanduser("~"))
        except IdentityError as error:
            raise RuntimeError(f"identidade invalida: {error}") from error

        manager = IntegrationManager(root)
        wrote_any = False
        for kind in ("agents", "skills"):
            _, plans = plan_deployments(kind, target_ids=["codex"], scope="project", identity_cfg=ident)
            statuses = manager.list(plans)
            to_write = [
                plan
                for plan, status in zip(plans, statuses)
                if status["state"] == "outdated" or (install_missing and status["state"] == "not-installed")
            ]
            if to_write:
                wrote_any = True
                manager.update(to_write)
        state = STATE_UPDATED if wrote_any else STATE_SKIPPED
        return {"id": "codex-project-agents", "state": state, "path": CODEX_PROJECT_AGENTS_DISPLAY_PATH}
    except Exception as error:
        return {
            "id": "codex-project-agents",
            "state": STATE_FAILED,
            "path": CODEX_PROJECT_AGENTS_DISPLAY_PATH,
            "message": str(error),
        }


def _resolve_project_targets(raw: str | None) -> list[str]:
    if not raw:
        return list(PROJECT_TARGET_IDS)
    requested = [value.strip() for value in raw.split(",") if value.strip()]
    unknown = [value for value in requested if value not in PROJECT_TARGET_IDS]
    if unknown:
        print(f"trackfw update: unknown target id(s): {', '.join(unknown)}")
        raise SystemExit(2)
    selected = set(requested)
    return [target_id for target_id in PROJECT_TARGET_IDS if target_id in selected]


def _copy_project_tree(src: str, dst: str) -> None:
    """Copies src into dst, skipping .git and node_modules, for use as a
    --dry-run sandbox that the real project tree is never written through."""
    def _ignore(directory: str, names: list[str]) -> set[str]:
        return {name for name in names if name in (".git", "node_modules")}

    for name in os.listdir(src):
        source_path = os.path.join(src, name)
        dest_path = os.path.join(dst, name)
        if name in (".git", "node_modules"):
            continue
        if os.path.isdir(source_path):
            shutil.copytree(source_path, dest_path, ignore=_ignore)
        else:
            shutil.copy2(source_path, dest_path)


def _run_project(args: argparse.Namespace) -> None:
    cwd = os.getcwd()
    if not os.path.exists(os.path.join(cwd, "trackfw.yaml")):
        print("Erro: trackfw.yaml não encontrado — execute trackfw init primeiro")
        raise SystemExit(1)

    target_ids = _resolve_project_targets(args.targets)
    dry_run = bool(args.dry_run)
    install_missing = bool(args.install_missing)

    apply_root = cwd
    tmp_dir = None
    if dry_run:
        tmp_dir = tempfile.mkdtemp(prefix="trackfw-update-")
        _copy_project_tree(cwd, tmp_dir)
        apply_root = tmp_dir

    try:
        from trackfw.generators.init_gen import inject_rules_detected
        from trackfw.generators.hooks import inject_hooks_detected

        targets: list[dict[str, Any]] = []
        for target_id in target_ids:
            if target_id == "agent-rules":
                targets.append(
                    _run_file_target(
                        "agent-rules",
                        ", ".join(AGENT_RULES_RELATIVE_PATHS),
                        apply_root,
                        AGENT_RULES_RELATIVE_PATHS,
                        inject_rules_detected,
                        dry_run,
                        install_missing,
                    )
                )
            elif target_id == "agent-hooks":
                targets.append(
                    _run_file_target(
                        "agent-hooks",
                        ", ".join(AGENT_HOOKS_RELATIVE_PATHS),
                        apply_root,
                        AGENT_HOOKS_RELATIVE_PATHS,
                        inject_hooks_detected,
                        dry_run,
                        install_missing,
                    )
                )
            elif target_id == "codex-project-agents":
                targets.append(_codex_project_agents_target(apply_root, dry_run, install_missing))
    finally:
        if tmp_dir:
            shutil.rmtree(tmp_dir, ignore_errors=True)

    summary = {
        STATE_UPDATED: sum(1 for target in targets if target["state"] == STATE_UPDATED),
        STATE_SKIPPED: sum(1 for target in targets if target["state"] == STATE_SKIPPED),
        STATE_MISSING: sum(1 for target in targets if target["state"] == STATE_MISSING),
        STATE_FAILED: sum(1 for target in targets if target["state"] == STATE_FAILED),
    }

    payload = {
        "scope": "project",
        "dry_run": dry_run,
        "targets": targets,
        "summary": summary,
    }

    if args.json:
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print("trackfw update — scope: project" + (" (dry-run)" if dry_run else ""))
        for target in targets:
            suffix = f" — {target['message']}" if target["state"] == STATE_FAILED and "message" in target else ""
            print(f"  {target['id']:<24} {target['state']:<8} ({target['path']}){suffix}")
        print(
            f"\nupdated={summary[STATE_UPDATED]} skipped={summary[STATE_SKIPPED]} "
            f"missing={summary[STATE_MISSING]} failed={summary[STATE_FAILED]}"
        )

    if summary[STATE_FAILED] > 0:
        raise SystemExit(1)

File: npm/src/commands/update.js
'use strict';

const { Command } = require('commander');
const fs = require('fs');
const os = require('os');
const path = require('path');
const identityStore = require('../identity');
const projectConfig = require('../config');
const { runFileTarget, validateTargets, buildDocument, humanReport, silenceConsole } = require('../lib/update-engine');

// `trackfw update` is project-scoped only — see docs/cli-parity.md,
// "`trackfw update` vs `trackfw update harness`". It must NEVER touch global
// state (~/.claude, ~/.codex, etc.). Global artifacts moved to
// `trackfw update harness` (update-harness.js).

// loadUpdateConfig replaces the former artisanal line-by-line scanner (readUpdateConfig) with
// the single config loader (../config, see ADR-2026-08-02-caminho-unico-de-leitura-do-trackfw-
// yaml-com-namespaces-tipados.md). Returns the same snake_case shape the rest of this file
// already consumes (cfg.hooks, cfg.ci, cfg.backend, cfg.frontend, cfg.pkg_manager) so downstream
// code (updateHooksSurgical, buildProjectTargets) needs no further change.
function loadUpdateConfig(rootDir) {
  const u = projectConfig.load(rootDir).update;
  return {
    hooks: u.hooks,
    ci: u.ci,
    backend: u.backend,
    frontend: u.frontend,
    pkg_manager: u.pkgManager,
  };
}

function updateHooksSurgical(cfg, rootDir) {
  const hooks = cfg.hooks || '';
  if (hooks === 'husky') {
    const hookPath = path.join(rootDir, '.husky', 'pre-commit');
    const content = fs.existsSync(hookPath) ? fs.readFileSync(hookPath, 'utf8') : '';
    if (content.includes('trackfw validate')) {
      console.log('  ✓ .husky/pre-commit — trackfw validate já presente');
    } else {
      fs.mkdirSync(path.join(rootDir, '.husky'), { recursive: true });
      fs.appendFileSync(hookPath, '\ntrackfw validate\n', 'utf8');
      try { fs.chmodSync(hookPath, 0o755); } catch (_) {}
      console.log('  ✓ .husky/pre-commit — trackfw validate injetado');
    }
  } else if (hooks === 'lefthook') {
    const lefthookPath = path.join(rootDir, 'lefthook.yml');
    const content = fs.existsSync(lefthookPath) ? fs.readFileSync(lefthookPath, 'utf8') : '';
    if (content.includes('trackfw-validate:') || content.includes('trackfw validate')) {
      console.log('  ✓ lefthook.yml — trackfw já presente');
    } else {
      fs.appendFileSync(lefthookPath, '\npre-commit:\n  commands:\n    trackfw-validate:\n      run: trackfw validate\n', 'utf8');
      console.log('  ✓ lefthook.yml — trackfw-validate injetado');
    }
  }
}

// codexProjectAgentsTarget — installs/reports the project-scoped Codex
// agents/skills bundle (catalog-based). Uses IntegrationManager.inspect,
// which is read-only, to compute state under --dry-run too; only the
// mutating manager.update() call is skipped when dryRun is true. Kept
// separate from runFileTarget because IntegrationManager already owns a
// correct, manifest-aware not-installed/current/outdated/modified state
// machine — re-deriving it via directory hashing would risk diverging from
// that source of truth (and would need the .trackfw manifest copied into
// the simulation, which is unnecessary complexity here).
function codexProjectAgentsTarget(cwd, identityConfig, { dryRun, installMissing }) {
  const id = 'codex-project-agents';
  const displayPath = '.codex/agents, .agents/skills';
  const detected = fs.existsSync(path.join(cwd, 'AGENTS.md')) || fs.existsSync(path.join(cwd, '.codex'));
  if (!detected) return { id, state: 'missing', path: displayPath };

  try {
    const { buildPlans, IntegrationManager } = require('../integrations');
    const manager = new IntegrationManager({ projectRoot: cwd });
    let wroteAny = false;
    for (const kind of ['agents', 'skills']) {
      const plans = buildPlans(kind, { targets: ['codex'], scope: 'project', identity: identityConfig });
      const statuses = manager.inspect(plans);
      const toWrite = plans.filter((_, index) => {
        const state = statuses[index].state;
        return state === 'outdated' || (installMissing && state === 'not-installed');
      });
      if (toWrite.length) {
        wroteAny = true;
        if (!dryRun) manager.update(toWrite);
      }
    }
    return { id, state: wroteAny ? 'updated' : 'skipped', path: displayPath };
  } catch (e) {
    return { id, state: 'failed', path: displayPath, message: e.message };
  }
}

// PROJECT_TARGET_IDS — the fixed declared order for `trackfw update`
// targets. `ci-workflow` and `git-hooks` only appear when the project's
// trackfw.yaml opted into a CI system / hook framework — see ambiguity
// note in the ML-6C handoff report about config-conditional target lists.
const PROJECT_TARGET_IDS = [
  'agent-rules',
  'agent-hooks',
  'codex-project-agents',
  'validate-script',
  'ci-workflow',
  'git-hooks',
  'claude-commands',
];

// buildProjectTargets — `wanted` (nullable) restricts which targets are
// even computed/applied. This must be enforced HERE, before any apply()
// runs, not as a post-hoc filter on the returned array: every target's
// apply() is a real filesystem side effect (outside --dry-run), so
// filtering afterwards would still have written every unrequested target.
function buildProjectTargets(cwd, cfg, identityConfig, { dryRun, installMissing }, wanted) {
  const generators = require('../generators/init');
  const discover = require('./discover');
  const hooksGen = require('../generators/hooks');
  const include = (id) => !wanted || wanted.includes(id);

  const targets = [];

  if (include('agent-rules')) targets.push(runFileTarget({
    id: 'agent-rules',
    path: 'CLAUDE.md, AGENTS.md, GEMINI.md, .github/copilot-instructions.md, .windsurfrules, .amazonq/developer/guidelines.md, .cursor/rules/trackfw.mdc',
    root: cwd,
    relPaths: ['CLAUDE.md', 'AGENTS.md', 'GEMINI.md', '.github/copilot-instructions.md', '.windsurfrules', '.amazonq/developer/guidelines.md', '.cursor/rules/trackfw.mdc'],
    apply: (root) => generators.injectRulesDetected(root),
    dryRun,
    installMissing,
  }));

  if (include('agent-hooks')) targets.push(runFileTarget({
    id: 'agent-hooks',
    path: '.claude/settings.json, .codex/hooks.json, .gemini/settings.json, .kiro/hooks/trackfw-attention.json, .github/hooks/trackfw-attention.json, .cursor/hooks.json, scripts/trackfw-attention-*.sh, scripts/trackfw-credential-guard.sh',
    root: cwd,
    relPaths: [
      '.claude/settings.json',
      '.codex/hooks.json',
      '.gemini/settings.json',
      '.kiro/hooks/trackfw-attention.json',
      '.github/hooks/trackfw-attention.json',
      '.cursor/hooks.json',
      'scripts/trackfw-attention-signal.sh',
      'scripts/trackfw-attention-cleanup.sh',
      'scripts/trackfw-credential-guard.sh',
    ],
    apply: (root) => {
      hooksGen.injectHooksDetected(root);
      hooksGen.generateAttentionScripts(cfg, root);
      hooksGen.generateCredentialGuardScript(root);
    },
    dryRun,
    installMissing,
  }));

  if (include('codex-project-agents')) targets.push(codexProjectAgentsTarget(cwd, identityConfig, { dryRun, installMissing }));

  if (include('validate-script')) targets.push(runFileTarget({
    id: 'validate-script',
    path: 'scripts/trackfw-validate.sh',
    root: cwd,
    relPaths: ['scripts/trackfw-validate.sh'],
    // Reuses generators/init.js's generateValidateScript — the SAME
    // generator `trackfw init` uses to write this file — not
    // discover.js's writeValidateScript, which produces a different
    // (simpler, non-per-backend) script and made every `update` re-run
    // report "updated" against a project actually already current
    // (ML-6H fix). loadUpdateConfig returns raw trackfw.yaml keys
    // (snake_case, e.g. "pkg_manager"); buildValidateScript expects the
    // camelCase shape used by the rest of the init generators.
    apply: (root) => generators.generateValidateScript({
      backend: cfg.backend,
      frontend: cfg.frontend,
      pkgManager: cfg.pkg_manager,
    }, root),
    dryRun,
    installMissing,
  }));

  if (include('ci-workflow') && (cfg.ci === 'github-actions' || cfg.ci === 'github_actions')) {
    targets.push(runFileTarget({
      id: 'ci-workflow',
      path: '.github/workflows/trackfw-validate.yml',
      root: cwd,
      relPaths: ['.github/workflows/trackfw-validate.yml'],
      apply: (root) => discover.writeCIWorkflowForce(root),
      dryRun,
      installMissing,
    }));
  }

  if (include('git-hooks') && (cfg.hooks === 'husky' || cfg.hooks === 'lefthook')) {
    const relPath = cfg.hooks === 'husky' ? '.husky/pre-commit' : 'lefthook.yml';
    targets.push(runFileTarget({
      id: 'git-hooks',
      path: relPath,
      root: cwd,
      relPaths: [relPath],
      apply: (root) => updateHooksSurgical(cfg, root),
      dryRun,
      installMissing,
    }));
  }

  if (include('claude-commands')) targets.push(runFileTarget({
    id: 'claude-commands',
    path: '.claude/commands/trackfw',
    root: cwd,
    relPaths: ['.claude/commands/trackfw'],
    apply: (root) => generators.generateClaudeCommandsForce(root),
    dryRun,
    installMissing,
  }));

  return targets;
}

const cmd = new Command('update');
cmd.description('Update trackfw-managed artifacts. Bare form is project-scoped (never touches global state); `update harness` updates the global harness instead.');
// `[mode]` is a plain positional argument, not a nested commander.Command —
// see the long comment in update-harness.js:run for why: a real subcommand
// redeclaring the same --json/--dry-run/--targets/--install-missing flags
// as this parent silently drops them (commander@12 quirk, confirmed by
// reproduction). One Command, one Option per flag, branch on `mode` inside
// a single action — this is the only structure that parses correctly.
cmd.argument('[mode]', 'Pass "harness" to update the global harness instead of the current project');
cmd.option('--dry-run', 'Compute and report states without writing anything');
cmd.option('--json', 'Emit the result document as JSON');
cmd.option('--targets <ids>', 'Comma-separated subset of target ids');
cmd.option('--install-missing', 'Allow missing targets to be installed');
cmd.action((mode, options) => {
  if (mode === 'harness') {
    return require('./update-harness').run(options);
  }
  if (mode) {
    console.error(`✗ Unknown update mode: ${mode} (expected "harness" or no argument)`);
    process.exit(1);
  }

  const cwd = process.cwd();
  const yaml = path.join(cwd, 'trackfw.yaml');
  if (!fs.existsSync(yaml)) {
    console.error('✗ trackfw.yaml não encontrado — execute trackfw init primeiro');
    process.exit(1);
  }

  let wanted;
  try {
    const requested = options.targets ? String(options.targets).split(',').map((s) => s.trim()).filter(Boolean) : [];
    wanted = validateTargets(PROJECT_TARGET_IDS, requested);
  } catch (e) {
    console.error(`✗ ${e.message}`);
    process.exit(1);
  }

  const cfg = loadUpdateConfig(cwd);
  const dryRun = Boolean(options.dryRun);
  const installMissing = Boolean(options.installMissing);

  // A identidade é carregada uma única vez, fora de qualquer try/catch —
  // um identity.json corrompido deve abortar o comando inteiro, nunca cair
  // silenciosamente para os nomes neutros default.
  const identityConfig = identityStore.load(os.homedir());

  // With --json, stdout must carry only the result document — apply()
  // functions log human progress lines as a side effect; silence them so a
  // consumer parsing --json output never has to skip preamble noise.
  const targets = options.json
    ? silenceConsole(() => buildProjectTargets(cwd, cfg, identityConfig, { dryRun, installMissing }, wanted))
    : buildProjectTargets(cwd, cfg, identityConfig, { dryRun, installMissing }, wanted);

  const doc = buildDocument('project', dryRun, targets);

  if (options.json) {
    console.log(JSON.stringify(doc, null, 2));
  } else {
    console.log(humanReport('project', dryRun, targets));
  }

  if (doc.summary.failed > 0) process.exitCode = 1;
});

module.exports = cmd;
